### PR TITLE
Added get group challenges route and initial tests

### PR DESCRIPTION
**Pull request by TheHollidayInn**
This address Get challenges for group and tests in #6509. I didn't see much logic here that needed to be abstracted, @paglias . Let me know if you disagree.

### CODE REVIEW

**Comment by crookedneighbor** [comment]
The `_.findIndex` syntax is a little clunky in the expectation. There's got to be a better way to write that expectation.

**Comment by TheHollidayInn** [comment]
Yea, I agree. I was trying to use the chai .include, but the object aren't equal. I'll look through the documentation a bit. Maybe, we can extend chai to use lodash.

**Comment by crookedneighbor** [comment]
The chai things plugin might be helpful. http://chaijs.com/plugins/chai-things

**Comment by KristianTashkov** [comment]
@blade when I suggested the same plugin you said you prefer not to have it :cry:

**Comment by KristianTashkov** [comment]
On a more constructive note, in such cases I do:

```
let challengeIds = challenges.map((challenge) => {
  return challenge._id;
});

expect(challengeIds).to.include(challenge._id);
```

**Comment by paglias** [comment]
@TheHollidayInn the PR looks good, a few things:
1. To keep consistency with how tasks are fetched I would rename api.getChallenges to api.getUserChallenges and use an url like `GET /challenges/user`.
2. I just realized both in getUserChallenges and getGroupChallenges we want to populate the leader and group data (the code is already there but commented). For populating the leader instead of hard-coding `profile.name` in the selected fields you can `import nameFields from '../../models/user';` as we're doing when populating the group leader

**Comment by paglias** [comment]
I just noticed we don't have tests for getUserChallenge either so it may be a good occasion to add them!

**Comment by paglias** [comment]
Actually don't add population, I'm not sure if it's really needed. We'll add it later if it becomes necessary

**Comment by crookedneighbor** [comment]
@KristianTashkov I thought you might say that. :smile:

I don't recall the exact details of why you wanted to use it, but I don't think the assertions looked quite as messy as these do. In any case, I think you were right, this would be a good plugin to add.

**Comment by TheHollidayInn** [comment]
So, Chai Things does not play well with Chai Keys: https://github.com/chaijs/chai-things/issues/24 .

File: test/api/v3/integration/challenges/GET-challenges_group_groupid.test.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      nonMember = await generateUser();
+
+      challenge = await generateChallenge(user, group);
+      challenge2 = await generateChallenge(user, group);
+    });
+
+    it('should return group challenges for non member', async () => {
+      let challenges = await nonMember.get(`/challenges/groups/${publicGuild._id}`);
+
+      expect(_.findIndex(challenges, {_id: challenge._id})).to.be.above(-1);
```
Comment (crookedneighbor) [inline test/api/v3/integration/challenges/GET-challenges_group_groupid.test.js]: If we can't use `chai-things`, I'd prefer this to be something like

``` js
let foundChallenge1 = _.find(challenges, { _id: challenge._id });
expect(foundChallenge1).to.exist;
```

File: website/src/controllers/api-v3/challenges.js [hunk 2]
```diff
@@ -121,7 +121,8 @@ api.joinChallenge = {
     let challenge = await Challenge.findOne({ _id: req.params.challengeId });
     if (!challenge) throw new NotFound(res.t('challengeNotFound'));
 
-    if (!challenge.hasAccess(user)) throw new NotFound(res.t('challengeNotFound'));
+    let group = await Group.getGroup({user, groupId: challenge.groupId, fields: '_id type privacy', optionalMembership: true});
```
Comment (paglias) [inline website/src/controllers/api-v3/challenges.js:124]: @TheHollidayInn why you added this check? `challenge.hasAccess` should haveve been enough

**Comment by paglias** [comment]
Except for my last comment it looks good!

File: website/src/controllers/api-v3/challenges.js (hunk 2 above)
Comment (KristianTashkov) [inline website/src/controllers/api-v3/challenges.js:124]: From what I undestand, you should be able to join challenges in public guilds without being a member of the guild and I guess this fixes it. But probably this logic should be put in `hasAccess`

Comment (paglias) [inline website/src/controllers/api-v3/challenges.js:124]: That's right, good catch

Comment (paglias) [inline website/src/controllers/api-v3/challenges.js:124]: And yes, this logic should be in hasAccess

Comment (TheHollidayInn) [inline website/src/controllers/api-v3/challenges.js:124]: @KristianTashkov Yep, this is exactly the case. 

I'll move the logic. :)

Comment (TheHollidayInn) [inline website/src/controllers/api-v3/challenges.js:124]: Actually, it seems the challenge.canView callas hasAccess, so basically, we only need to check viewAccess.

**Comment by TheHollidayInn** [comment]
Updated!

**Comment by paglias** [comment]
@TheHollidayInn thanks!

in the end i preferred to use a modified hasAccess that will work with public guilds too so it's almost the same as canView except that it doesn't return true if you're already a member of the challenge because that's handled separately by challenge.isMember and in case will return a `userAlreadyInChallenge` error.

**Comment by TheHollidayInn** [comment]
Sounds good, thanks!

**Pull request merged**